### PR TITLE
ZeroDivisionError when groupby rank with method="dense" and pct=True

### LINKED ISSUE #23666: ZeroDivisionError when groupby rank with method="dense" and pct=True
When I tried to use groupby rank function with method="dense", pct=True options, I encountered the ZeroDivisionError. 

#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

df = pd.DataFrame({"A": [1, 1, 1, 2, 2, 2],
                   "B": [1, 1, 1, 1, 2, 2],
                   "C": [1, 2, 1, 1, 1, 2]})
df.groupby(["A", "B"])["C"].rank(method="dense", pct=True)
```
error:

```
Traceback (most recent call last):
  File "c:/Users/<user_name>/Documents/test.py", line 6, in <module>
    df.groupby(["A", "B"])["C"].rank(method="dense", pct=True)
  File "C:\Users\<user_name>\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 1906, in rank
    na_option=na_option, pct=pct, axis=axis)
  File "C:\Users\<user_name>\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 1025, in _cython_transform
    **kwargs)
  File "C:\Users\<user_name>\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 2630, in transform
    return self._cython_operation('transform', values, how, axis, **kwargs)
  File "C:\Users\<user_name>\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 2590, in _cython_operation
    **kwargs)
  File "C:\Users\<user_name>\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 2664, in _transform
    transform_func(result, values, comp_ids, is_datetimelike, **kwargs)
  File "C:\Users\<user_name>\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 2479, in wrapper
    return f(afunc, *args, **kwargs)
  File "C:\Users\<user_name>\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 2431, in <lambda>
    kwargs.get('na_option', 'keep')
  File "pandas\_libs\groupby_helper.pxi", line 1292, in pandas._libs.groupby.group_rank_int64
ZeroDivisionError: float division
```

#### Problem description

I encountered ZeroDivisionError when I tried to use the groupby rank function. 

I can't find out exactly what a problem is.  But when I drop either method="dense" or pct=True option, the above code works.

If some elements in the above DataFrame are changed, this error disappear.  For example, the following code gives the expected output.

```python
df = pd.DataFrame({"A": [1, 1, 1, 2, 2, 2],
                   "B": [1, 1, 1, 1, 2, 2],
                   "C": [1, 2, 1, 0, 1, 2]}) # a little change in column C
df.groupby(["A", "B"])["C"].rank(method="dense", pct=True)
```

output:
```
0    0.5
1    1.0
2    0.5
3    1.0
4    0.5
5    1.0
Name: C, dtype: float64
```

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.6.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.4
pytest: 3.5.1
pip: 10.0.1
setuptools: 39.1.0
Cython: 0.28.2
numpy: 1.14.5
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.4.0
sphinx: 1.7.4
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: 1.2.1
tables: 3.4.3
numexpr: 2.6.5
feather: None
matplotlib: 3.0.0
openpyxl: 2.5.3
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.4
lxml: 4.2.5
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.7
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by Koustav-Samaddar**
- [x] closes #23666 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
the first thing to always write is a test

File: pandas/_libs/groupby_helper.pxi.in [hunk 1]
```diff
@@ -588,7 +588,16 @@ def group_rank_{{name}}(ndarray[float64_t, ndim=2] out,
                 if out[i, 0] != out[i, 0] or out[i, 0] == NAN:
                     out[i, 0] = NAN
                 else:
-                    out[i, 0] = out[i, 0] / grp_sizes[i, 0]
+                    with gil:
```
Comment (WillAyd) [inline pandas/_libs/groupby_helper.pxi.in]: Instead of going back to the gil can't you just check if the denominator is zero?

File: pandas/tests/groupby/test_rank.py [hunk 2]
```diff
@@ -290,3 +290,17 @@ def test_rank_empty_group():
     result = df.groupby(column).rank(pct=True)
     expected = DataFrame({"B": [0.5, np.nan, 1.0]})
     tm.assert_frame_equal(result, expected)
+
+
+def test_rank_zero_div():
+    # GH 23666
+    df = pd.DataFrame({
+        "A": [1, 1, 1, 2, 2, 2],
```
Comment (WillAyd) [inline pandas/tests/groupby/test_rank.py]: Can you make this test case any smaller? As minimal of an example to produce the error as possible would be clearest

File: doc/source/whatsnew/v0.24.0.rst [hunk 3]
```diff
@@ -1420,6 +1420,7 @@ Groupby/Resample/Rolling
 - Bug in :meth:`DataFrame.expanding` in which the ``axis`` argument was not being respected during aggregations (:issue:`23372`)
 - Bug in :meth:`pandas.core.groupby.DataFrameGroupBy.transform` which caused missing values when the input function can accept a :class:`DataFrame` but renames it (:issue:`23455`).
 - Bug in :func:`pandas.core.groupby.GroupBy.nth` where column order was not always preserved (:issue:`20760`)
+- Bug in :meth:`pandas.core.groupby.DataFrameGroupBy.rank` with ``method='dense'`` and ``pct=True`` when a group has only one member would raise a ZeroDivisionError (:issue:`23666`).
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.24.0.rst]: Can put the literal of `ZeroDivisionError` in double backticks

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_rank.py [hunk 4]
```diff
@@ -290,3 +290,17 @@ def test_rank_empty_group():
...
+        "B": [1, 1, 1, 1, 2, 2],
+        "C": [1, 2, 1, 1, 1, 2]
+    })
+
+    try:
```
Comment (WillAyd) [inline pandas/tests/groupby/test_rank.py]: You don't need a try except here - exceptions will be propagated naturally and cause a test failure

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_rank.py (hunk 4 above)
Comment (WillAyd) [reply pandas/tests/groupby/test_rank.py]: Would rather just be explicit about the expected result and test for that

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/_libs/groupby_helper.pxi.in [hunk 5]
```diff
@@ -588,7 +588,13 @@ def group_rank_{{name}}(ndarray[float64_t, ndim=2] out,
                 if out[i, 0] != out[i, 0] or out[i, 0] == NAN:
                     out[i, 0] = NAN
                 else:
-                    out[i, 0] = out[i, 0] / grp_sizes[i, 0]
```
Comment (jreback) [inline pandas/_libs/groupby_helper.pxi.in]: you need to fill with NAN if the grp_sizes[i, 0] == 0

Comment (jreback) [reply pandas/_libs/groupby_helper.pxi.in]: make this another elif condition, then the else can be the division.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/groupby_helper.pxi.in (hunk 5 above)
Comment (Koustav-Samaddar) [reply pandas/_libs/groupby_helper.pxi.in]: Aren't lines 588-589 taking care of the NAN edge case?

In my testing I have only ever been able to get grp_size[i, 0] == 0 in the bugged case (pct=True, method="dense").

So my idea of the fix was that grp_size[i, 0] should never be 0 in the else block and if it is, then it should be 1 (not modifying out[i, 0] would be the same as div by 1).

Comment (jreback) [reply pandas/_libs/groupby_helper.pxi.in]: no, you need to add this case there as well, if the denominator is 0, the result is NaN

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/groupby_helper.pxi.in (hunk 5 above)
Comment (Koustav-Samaddar) [reply pandas/_libs/groupby_helper.pxi.in]: As it is right now, doing that would cause incorrect results (would fail the test I added in test_rank.py). The reasoning behind this is:

1. grp_size[i, 0] should legitimately be 0 only if all the values in a group are NaN. In this case the if statement in Lines 588-589 takes care of this case.

2. grp_size[i, 0] shouldn't be 0 but is in the case where there is a group with only one member with a non NaN value but another index also has the same value but in a different group && method="density". In this case the correct grp_size[i, 0] should be 1 and it being 0 is what causes BUG 23666.

Therefore my fix is only performing the division by grp_size[i, 0] when it is not equal to 0 in the if block [Line 591] because the else condition only demands division by 1 (the corrected grp_size[i, 0]).

To summarise, making the result NaN on grp_size[i, 0] would stop the ZeroDivisionError from occurring in BUG 23666 and replace it with incorrect behaviour.

An additional note: the conclusion of my testing is that there is a flaw in the calculation of grp_sizes[i, 0] itself under these circumstances, but in my testing I have uncovered that there are other parameters (not just method="dense") that can have weird behaviour that is caused due to incorrect grp_sizes[i, 0]. Once, I'm able to determine the extent of this I'm going to create a new issue that would target the fixing of grp_size[i, 0] calculation bug. This fix is a stop-gap measure till then since this would correct one instance of incorrect rank behaviour.

Comment (WillAyd) [reply pandas/_libs/groupby_helper.pxi.in]: Please improve the test case to make an assertion about the result of this expression. It’s not enough to assert that it simply computes which is the behavior now. 

Doing that would aid this discussion

Comment (Koustav-Samaddar) [reply pandas/_libs/groupby_helper.pxi.in]: Ah sorry, my bad. I forgot to push it.

File: pandas/_libs/groupby_helper.pxi.in [hunk 6]
```diff
@@ -588,7 +588,13 @@ def group_rank_{{name}}(ndarray[float64_t, ndim=2] out,
...
+                    if grp_sizes[i, 0] != 0:
+                        # BUG 23666 - tie-breaking for dense causes the
```
Comment (WillAyd) [inline pandas/_libs/groupby_helper.pxi.in]: I think this comment would be better served if shortened and placed before the condition rather than within

**Review by WillAyd — CHANGES_REQUESTED** [review]
Minor nit on comment placement but otherwise lgtm for this change. If you find a way to refactor as you keep looking at it certainly open to subsequent PRs

**Comment by WillAyd** [comment]
Also make sure you check the Travis failures. Think you have a styling issue somewhere

**Comment by Koustav-Samaddar** [comment]
> Also make sure you check the Travis failures. Think you have a styling issue somewhere

I skimmed through the Travis logs and wasn't able to find (/ didn't understand) what the issue was. I didn't modify any of the source files that I saw were mentioned in the logs. There was an error/issue raised by scripts/tests/test_validate_docstrings.py but I couldn't understand the logs from it.

I rechecked with pep8 and flake8 and couldn't find anything sticking out.

**Comment by WillAyd** [comment]
The failure is in `ci/code_checks.sh` if you run locally should give you visibility into issue.

@datapythonista FYI failures in that script don't provide a useful error message in CI currently. Not sure if that's on your radar

File: pandas/tests/groupby/test_rank.py [hunk 7]
```diff
@@ -290,3 +290,15 @@ def test_rank_empty_group():
...
+def test_rank_zero_div():
+    # GH 23666
+    df = DataFrame({
+        "A": [1, 2],
+        "B": [1, 1]
+    })
+
+    result = df.groupby("A").rank(pct=True, method="dense")
+    expected = DataFrame({"B": [1.0, 1.0]})
+    tm.assert_frame_equal(result, expected)
```
Comment (jreback) [inline pandas/tests/groupby/test_rank.py]: this is a lint error

File: pandas/tests/groupby/test_rank.py [hunk 8]
```diff
@@ -290,3 +290,15 @@ def test_rank_empty_group():
     result = df.groupby(column).rank(pct=True)
     expected = DataFrame({"B": [0.5, np.nan, 1.0]})
     tm.assert_frame_equal(result, expected)
+
+
+def test_rank_zero_div():
+    # GH 23666
+    df = DataFrame({
```
Comment (jreback) [inline pandas/tests/groupby/test_rank.py]: can you add a case which has a NaN, one which has multiple values in a group, and one which has NO values in the group.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by datapythonista** [comment]
@WillAyd the error message `./pandas/tests/groupby/test_rank.py:304:44: W292 no newline at end of file` is coming from `flake8`.

If you mean that they are difficult to find, that's addressed in #22854, which should make them much easier and quicker to find. We have a permissions problem with Azure, that I expect to be fixed early next week, so with some luck we have it merged very soon.

**Comment by Koustav-Samaddar** [comment]
I ran `LINT=true ci/code_checks.sh` on my machine and it didn't fail anything. However, on code submission, Travis is still failing.

**Comment by datapythonista** [comment]
The travis failure was unrelated to the linting, it was caused by a url that could not be fetched. I rerun it let's see if it succeeds now.

**Comment by jreback** [comment]
> grp_size[i, 0] shouldn't be 0 but is in the case where there is a group with only one member with a non NaN value and another index also has the same value but in a different group && method="density". In this case the correct grp_size[i, 0] should be 1 and it being 0 is what causes BUG 23666.

I don't buy this. why would a grp_size *ever* be 0?

**Comment by Koustav-Samaddar** [comment]
> > grp_size[i, 0] shouldn't be 0 but is in the case where there is a group with only one member with a non NaN value and another index also has the same value but in a different group && method="density". In this case the correct grp_size[i, 0] should be 1 and it being 0 is what causes BUG 23666.
> 
> I don't buy this. why would a grp_size _ever_ be 0?

If I have a group where all the values are `NaN`s, then `grp_size[i, 0]` should be zero. However, the `if out[i, 0] == NaN` block makes it so that all the members (not sure if right term) of such a group get caught there, and never make it past to the check of `grp_size[i, 0] == 0`.

TL;DR there are valid cases where `grp_size[i, 0] == 0` but they will never enter the condition that checks `if grp_size[i, 0] == 0`

**Comment by Koustav-Samaddar** [comment]
When I run ci/code_checks.sh on my local machine it doesn't raise any errors. However, in Travis I'm getting `The command "ci/code_checks.sh" exited with 1.`. I still am unable to ascertain from the Travis logs what's going wrong/failing.

Any help is greatly appreciated!

**Comment by jreback** [comment]
> there are valid cases where grp_size[i, 0] == 0 

and what is an example of this
don’t reference the existing code
rather show an input an output that has no group size yet has a result

**Comment by jreback** [comment]
and what would the result frame look like

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_rank.py [hunk 9]
```diff
@@ -290,3 +290,18 @@ def test_rank_empty_group():
...
     tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("df_dicts", [
+    ({"A": [1, 2], "B": [1, 1]}, {"B": [1.0, 1.0]}),
+    ({"A": [1, 1, 2, 2], "B": [1, 2, 1, 2]}, {"B": [0.5, 1.0, 0.5, 1.0]}),
+    ({"A": [1, 1, 2, 2], "B": [1, 2, 1, np.nan]}, {"B": [0.5, 1.0, 1.0, np.nan]}),
+    ({"A": [1, 1, 2], "B": [1, 2, np.nan]}, {"B": [0.5, 1.0, np.nan]})
+])
+def test_rank_zero_div(df_dicts):
```
[the remark this answers is not in the document]
Comment (Koustav-Samaddar) [reply pandas/tests/groupby/test_rank.py]: I'm not sure if I understand this correctly, so I followed the same styling as a previous test in the same file.

I hope the changes are more readable!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @WillAyd over to you.

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @Koustav-Samaddar !
